src/RawPOSIX/src/safeposix/syscalls/fs_calls.rs

```diff
@@ -370,6 +370,187 @@ impl Cage {
+        if cwd_ptr.is_null() {
+            let errno = get_errno();
+            return handle_errno(errno, "getcwd");
+        }
+
+        let pwd = unsafe { CStr::from_ptr(cwd_buf.as_ptr() as *const c_char) }
+            .to_str()
+            .unwrap();
+
+        // Adjust the result to user perspective
+        // Verify if libcbuf_str starts with the current working directory (pwd)
+        let adjusted_result = if libcbuf_str.starts_with(pwd) {
+            libcbuf_str.to_string()
+        } else {
+            format!("{}/{}", pwd, libcbuf_str)
+        };
+        let new_root = format!("{}/", LIND_ROOT);
+        let final_result = adjusted_result.strip_prefix(&new_root).unwrap_or(&adjusted_result);
+
+        // Check the length and copy the appropriate amount of data to buf
+        let bytes_to_copy = std::cmp::min(buflen, final_result.len());
+        unsafe {
+            std::ptr::copy_nonoverlapping(final_result.as_ptr(), buf, bytes_to_copy);
+        }
+
+        bytes_to_copy as i32
+    }
+    
+    /* 
+    * The readlinkat syscall builds upon the readlink syscall, with additional handling for the provided fd. 
+    * There are two main cases to consider:
+    *
+    * When fd is the special value AT_FDCWD:
+    * In this case, we first retrieve the current working directory path. We then append the user-provided path 
+    * to this directory path to create a complete path. After this, the handling is identical to the readlink 
+    * syscall. Therefore, the implementation delegates the underlying work to the readlink syscall.
+    *
+    * One notable point is that when fd = AT_FDCWD, there is no need to convert the virtual fd. Due to Rust's 
+    * variable scoping rules and for safety considerations (we must use the predefined fdtable API). This results 
+    * in approximately four lines of repetitive code during the path conversion step. If we plan to optimize 
+    * the implementation in the future, we can consider abstracting this step into a reusable function to avoid 
+    * redundancy.
+    *
+    * When fd is a directory fd:
+    * Handling this case is difficult without access to kernel-space code. In Linux, there is no syscall that 
+    * provides a method to resolve the directory path corresponding to a given dirfd. The Linux kernel handles 
+    * this step by utilizing its internal dentry data structure, which is not accessible from user space. 
+    * Therefore, in the RawPOSIX implementation, we assume that all paths are absolute to simplify the resolution 
+    * process.
+    *
+    */
+    pub fn readlinkat_syscall(
+        &self, 
+        virtual_fd: i32, 
+        path: &str, 
+        buf: *mut u8, 
+        buflen: usize,
+    ) -> i32 {
+        let mut libcret;
+        let mut path = path.to_string();
+        let libc_buflen = buflen + LIND_ROOT.len();
+        let mut libc_buf = vec![0u8; libc_buflen];
+        if virtual_fd == libc::AT_FDCWD {
```

Comment:
we moved to using our own defined constants instead of libc, but similar to qianxi's comment were waiting on the memory PR to be merged. So this will have to be fixed depending what ordering we merge.